Sources/Datadog/Core/System/NetworkConnectionInfoProvider.swift

```diff
@@ -40,90 +40,40 @@ internal protocol NetworkConnectionInfoProviderType {
 }
 
 internal class NetworkConnectionInfoProvider: NetworkConnectionInfoProviderType {
-    private let queue = DispatchQueue.global(qos: .utility)
-    private let monitor: NWCurrentPathMonitor
+    private let fetchBlock: () -> NetworkConnectionInfo
+    private let cancelBlock: (() -> Void)?
 
-    init(monitor: NWCurrentPathMonitor = NWPathMonitor()) {
-        self.monitor = monitor
-        self.monitor.start(queue: queue)
+    var current: NetworkConnectionInfo {
+        return fetchBlock()
     }
 
-    deinit {
-        monitor.cancel()
+    convenience init() {
+        if #available(iOS 12, *) {
+            self.init(NWPathMonitor())
+        } else {
+            self.init(iOS11PathMonitor())
+        }
     }
 
-    var current: NetworkConnectionInfo {
-        let currentPath = monitor.currentPathInfo()
-        let availableInterfaces: [NetworkConnectionInfo.Interface] = {
-            currentPath.availableInterfaceTypes.map { interfaceType in
-                switch interfaceType {
-                case .wifi: return .wifi
-                case .wiredEthernet: return .wiredEthernet
-                case .cellular: return .cellular
-                case .loopback: return .loopback
-                case .other: return .other
-                @unknown default: return .other
-                }
-            }
-        }()
-
-        let reachability: NetworkConnectionInfo.Reachability = {
-            switch currentPath.status {
-            case .satisfied: return .yes
-            case .requiresConnection: return .maybe
-            case .unsatisfied: return .no
-            @unknown default: return .maybe
-            }
-        }()
-
-        return NetworkConnectionInfo(
-            reachability: reachability,
-            availableInterfaces: availableInterfaces,
-            supportsIPv4: currentPath.supportsIPv4,
-            supportsIPv6: currentPath.supportsIPv6,
-            isExpensive: currentPath.isExpensive,
-            isConstrained: currentPath.isConstrained
-        )
+    @available(iOS 12, *)
+    init(_ provider: NWPathMonitor) {
+        self.fetchBlock = {
+            return provider.current
+        }
+        self.cancelBlock = { [weak provider] in
+            provider?.cancel()
+        }
+        provider.start(queue: DispatchQueue.global(qos: .utility))
```

Comment:
I'm still unsure about this inlined queue.

First, the [`.start(queue:)` documentation](https://developer.apple.com/documentation/network/nwpathmonitor/2998737-start) doesn't say anything about ownership of this queue so we might be badly surprised in next version of iOS and / or when different build optimizations are applied by SDK users.

Second, following ["2) Always start with serial queues, and only use concurrent queues as a last resort"](https://www.fadel.io/blog/posts/ios-performance-tips-you-probably-didnt-know/) it's IMO safer to start with our custom serial `DispatchQueue(label: "com.datadoghq.network-connection-info")` queue than choosing concurrent global queue.

Reply:
yes, that sounds safer 👍 

[Apple doc shows another way](https://developer.apple.com/documentation/dispatch/dispatchqueue):

> For serial tasks, set the target of your serial queue to one of the global concurrent queues. That way, you can maintain the serialized behavior of the queue while minimizing the number of separate queues creating threads.

i'm thinking of following that way